draft-ietf-quic-transport.md

```diff
@@ -1829,10 +1829,21 @@ tokens that would be accepted by the server.  Only the server requires access to
 the integrity protection key for tokens.
 
 There is no need for a single well-defined format for the token because the
-server that generates the token also consumes it.  A token could include
-information about the claimed client address (IP and port), a timestamp, and any
-other supplementary information the server will need to validate the token in
-the future.
+server that generates the token also consumes it.  Tokens sent in Retry packets
+SHOULD include information that allows the server to verify that the source IP
+address and port in client packets remains constant.
+
+Servers might use tokens from NEW_TOKEN in deciding not to send a Retry packet,
+even if the client address has changed.  A token that was provided in
+NEW_TOKEN cannot be used for address validation if the client address is not the
+same, though servers MAY allow for the possibility of changes arising from new
+mappings at a NAT.
```

Comment:
Would it make sense to change "though..." to something like: "unless the server can assume that the address change happened within the same network (e.g., NAT rebinding)."?

The reason I suggest this change is because client address might change due to reasons other than NAT, for example, DHCP or SLAAC.

Reply:
I think that new mappings at a NAT are all I really want to allow for here, recognizing that this is likely indistinguishable from a new address from SLAAC or a DHCP lease.

We shouldn't really want a server to spray packets at unwitting hosts within the network just because another one of those hosts has a valid token, more so for large carrier-grade NATs.  But nor can we realistically prevent that, so this needs to be a little squishy.

Reply:
Thank you for the response. To me it seems that what we are actually looking for is an upper-case SHOULD that limits the scope of NEW_TOKEN tokens to one IP address. IMO that is the basis that we need to state. In addition, we can talk about when the scope can be wider, but that is less important.

I am not sure where the upper-case SHOULD should go into. It could be this section or [section 8.1.3](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-address-validation-for-futu). But assuming that it is the former, maybe we can replace this paragraph with something as simple as: "Tokens sent in NEW_TOKEN frames SHOULD include information that allows the server to verify if the client address is stable."

WDYT?

Reply:
(Oops, missed this comment.)

I didn't want to recommend that the IP address remain constant as NAT does not always limit itself to port changes over time.  However, I suspect that the right answer in that case is to not assume that the address is valid.  NATs that use multiple addresses often serve many clients, so the risk of amplification is heightened.

That said, I do like the brevity of what you have.  I'll use that, but I do want to keep the Retry sentence (that's important).